File: crates/examples3d/joint_ball3.rs
use khal::backend::GpuTimestamps;
use nexus_viewer3d::NexusViewer;
use nexus3d::prelude::{NexusCapacities, NexusPipeline, NexusState};
use rapier3d::prelude::*;

pub async fn run(
    viewer: &mut NexusViewer,
    pipeline: &mut NexusPipeline,
) -> anyhow::Result<NexusState> {
    /*
     * World
     */
    let capacities = NexusCapacities::default().rbd_collisions(350_000);
    let mut state = NexusState::new(capacities);

    let rad = 0.4;
    let ni = 200;
    let nk = 301;
    let shift = 1.0;
    let center = Vec3::new(nk as f32 * shift / 2.0, 0.0, ni as f32 * shift / 2.0);

    let mut body_handles = Vec::new();

    // A lot of joints. Kind of like a piece of cloth.
    for k in 0..nk {
        for i in 0..ni {
            let fk = k as f32;
            let fi = i as f32;

            let status = if ((i == 0 || i == ni - 1) && (k % 4 == 0 || k == ni - 1))
                || ((k == 0 || k == nk - 1) && (i % 4 == 0 || i == nk - 1))
            {
                RigidBodyType::Fixed
            } else {
                RigidBodyType::Dynamic
            };

            let rigid_body = RigidBodyBuilder::new(status)
                .translation(Vec3::new(fk * shift, 0.0, fi * shift) - center)
                .build();
            let collider = if status == RigidBodyType::Fixed {
                ColliderBuilder::cuboid(rad, rad, rad).build()
            } else {
                ColliderBuilder::ball(rad).density(10.0).build()
            };
            let shape = collider.shared_shape().clone();
            let child_handle = state.insert_rigid_body(rigid_body, collider);
            viewer.insert_shape(child_handle, &shape, Pose::IDENTITY);

            // Vertical joint.
            if i > 0 {
                let parent_handle = *body_handles.last().unwrap();
                let joint = SphericalJointBuilder::new().local_anchor2(Vec3::new(0.0, 0.0, -shift));
                state.insert_impulse_joint(parent_handle, child_handle, joint);
            }

            // Horizontal joint.
            if k > 0 {
                let parent_index = body_handles.len() - ni;
                let parent_handle = body_handles[parent_index];
                let joint = SphericalJointBuilder::new().local_anchor2(Vec3::new(-shift, 0.0, 0.0));
                state.insert_impulse_joint(parent_handle, child_handle, joint);
            }

            body_handles.push(child_handle);
        }
    }

    // Some rigid-bodies to fall on top.
    let nj = 10;
    let nk = nk / 3;
    let ni = ni / 6;
    let rad = rad * 2.5;

    for k in 0..nk {
        for i in 0..ni {
            for j in 0..nj {
                let body = RigidBodyBuilder::dynamic()
                    .translation(Vec3::new(
                        (k as f32 - nk as f32 / 2.0) * rad * 2.1,
                        j as f32 * rad * 2.1 + 2.0,
                        (i as f32 - ni as f32 / 2.0) * rad * 2.1,
                    ))
                    .build();
                let collider = ColliderBuilder::cuboid(rad, rad, rad).build();
                let shape = collider.shared_shape().clone();
                let handle = state.insert_rigid_body(body, collider);
                viewer.insert_shape(handle, &shape, Pose::IDENTITY);
            }
        }
    }

    // Optional, useful so we can render even before starting the simulation.
    let mut timestamps = GpuTimestamps::new(viewer.backend(), 2048);
    viewer
        .scene3d_mut()
        .add_directional_light(glamx::Vec3::new(1.0, -2.0, 3.0));
    state.finalize(viewer.backend()).await?;

    while viewer.render_frame().await {
        if viewer.simulating() {
            pipeline
                .simulate(viewer.backend(), &mut state, Some(&mut timestamps))
                .await?;
        }
        viewer.sync(&mut state, Some(&mut timestamps)).await?;
    }

    Ok(state)
}

File: crates/nexus_python3d/examples/joint_ball3.py
"""Python port of `crates/examples3d/joint_ball3.rs`.

A large grid of spherical (ball) joints, like a piece of cloth, with a
pile of rigid bodies falling on top.
"""

from nexus3d import (
    NexusViewer,
    NexusPipeline,
    NexusState,
    RigidBodyBuilder,
    ColliderBuilder,
    SphericalJointBuilder,
    GpuTimestamps,
    Vec3,
    Pose,
)


def run(viewer: NexusViewer, pipeline: NexusPipeline) -> NexusState:
    state = NexusState()

    rad = 0.4
    ni = 200
    nk = 301
    shift = 1.0
    center = Vec3(nk * shift / 2.0, 0.0, ni * shift / 2.0)

    body_handles = []

    # A lot of joints. Kind of like a piece of cloth.
    for k in range(nk):
        for i in range(ni):
            fk = float(k)
            fi = float(i)

            fixed = ((i == 0 or i == ni - 1) and (k % 4 == 0 or k == ni - 1)) or (
                (k == 0 or k == nk - 1) and (i % 4 == 0 or i == nk - 1)
            )

            if fixed:
                rigid_body = (
                    RigidBodyBuilder.fixed()
                    .translation(Vec3(fk * shift, 0.0, fi * shift) - center)
                    .build()
                )
                collider = ColliderBuilder.cuboid(rad, rad, rad).build()
            else:
                rigid_body = (
                    RigidBodyBuilder.dynamic()
                    .translation(Vec3(fk * shift, 0.0, fi * shift) - center)
                    .build()
                )
                collider = ColliderBuilder.ball(rad).density(10.0).build()

            shape = collider.shared_shape()
            child_handle = state.insert_rigid_body(rigid_body, collider)
            viewer.insert_shape(child_handle, shape, Pose.IDENTITY)

            # Vertical joint.
            if i > 0:
                parent_handle = body_handles[-1]
                joint = SphericalJointBuilder.new().local_anchor2(
                    Vec3(0.0, 0.0, -shift)
                )
                state.insert_impulse_joint(parent_handle, child_handle, joint)

            # Horizontal joint.
            if k > 0:
                parent_index = len(body_handles) - ni
                parent_handle = body_handles[parent_index]
                joint = SphericalJointBuilder.new().local_anchor2(
                    Vec3(-shift, 0.0, 0.0)
                )
                state.insert_impulse_joint(parent_handle, child_handle, joint)

            body_handles.append(child_handle)

    # Some rigid-bodies to fall on top.
    nj = 10
    nk = nk // 3
    ni = ni // 6
    rad = rad * 2.5

    for k in range(nk):
        for i in range(ni):
            for j in range(nj):
                body = (
                    RigidBodyBuilder.dynamic()
                    .translation(
                        Vec3(
                            (k - nk / 2.0) * rad * 2.1,
                            j * rad * 2.1 + 2.0,
                            (i - ni / 2.0) * rad * 2.1,
                        )
                    )
                    .build()
                )
                collider = ColliderBuilder.cuboid(rad, rad, rad).build()
                shape = collider.shared_shape()
                handle = state.insert_rigid_body(body, collider)
                viewer.insert_shape(handle, shape, Pose.IDENTITY)

    timestamps = GpuTimestamps(viewer, 2048)
    viewer.add_directional_light(Vec3(1.0, -2.0, 3.0))
    state.finalize(viewer)

    while viewer.render_frame():
        if viewer.simulating():
            pipeline.simulate(viewer, state, timestamps)
        viewer.sync(state, timestamps)

    return state


def main() -> None:
    viewer = NexusViewer()
    viewer.init_backend()
    pipeline = NexusPipeline()
    pipeline.preload_pipelines(viewer)
    run(viewer, pipeline)


if __name__ == "__main__":
    main()
    import os

    os._exit(0)
